Fix/submission error dialog

AndroidMedaGalaxy opened the pull request:
<!--
Thank you for supporting us with your Pull Request! 🙌 ❤️
Before submitting, please take the time to check the points below and provide some descriptive information.
-->

## Checklist

* [X] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [x] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [x] Set a speaking title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes # 41)
* [ ] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) - Pull Requests that are not linked to an issue with you as an assignee will be closed, except for minor fixes for typos or grammar mistakes in *documentation or code comments*.
* [x] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [x] Make sure that your PR is not introducing _unnecessary_ reformatting (e.g., introduced by on-save hooks in your IDE)
* [x] Make sure that your PR does not contain changes in text files, therefore the PR must not contain changes in values/strings/* and / or assets/* (see issue #332 for further information)
* [X] Make sure that your PR does not contain compiled sources (already set by the default .gitignore) and / or binary files

## Description
<!-- Please be brief in describing which issue is solved by your PR or which enhancement it brings -->

AndroidMedaGalaxy committed:
submission error dialog

AndroidMedaGalaxy committed:
submission_error_dialog_german_strings_added

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

harambasicluka committed:
Merge branch 'dev' into fix/submission_error_dialog

harambasicluka commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragment.kt:
```diff
@@ -6,6 +6,7 @@ import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
 import android.view.accessibility.AccessibilityEvent
+import androidx.appcompat.app.AlertDialog
```
this import seems to be unused

ritsam replied:
Hi Luka,

I am still making few changes in JUnit test. Please do not merge the code just yet.

ritsam replied:
It's been taken care of

Fabian-K commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragment.kt:
```diff
@@ -178,6 +188,34 @@ class SubmissionResultPositiveOtherWarningFragment : Fragment(),
     override fun onKeySharePermissionGranted(keys: List<TemporaryExposureKey>) {
         super.onKeySharePermissionGranted(keys)
         submissionViewModel.submitDiagnosisKeys(keys)
+        if (submissionViewModel.submissionState != ApiRequestState.SUCCESS) {
+            showSubmitErrorDialog()
+        }
```
I think this would show the "no diagnosis key" dialog for all sorts of failures and not only when no keys are available.

Also, error handling is already in onViewCreated -> submissionViewModel.submissionError.observeEvent. I think we should avoid having it at two places.

ritsam replied:
Made the necessary changes

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
submission error dialog unit test added

AndroidMedaGalaxy committed:
pipeline related changes

Fabian-K commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragment.kt:
```diff
@@ -177,7 +187,31 @@ class SubmissionResultPositiveOtherWarningFragment : Fragment(),
+     * @see SubmissionViewModel.submissionState
+     * @see AlertDialog.setCancelable
+     */
+    private fun showSubmitErrorDialog() {
```
I´d suggest to use a more descriptive name for this. It´s not a general error dialog but specific for one special case

ritsam replied:
refracted the name

Fabian-K commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragment.kt:
```diff
@@ -177,7 +187,31 @@ class SubmissionResultPositiveOtherWarningFragment : Fragment(),
     override fun onKeySharePermissionGranted(keys: List<TemporaryExposureKey>) {
         super.onKeySharePermissionGranted(keys)
-        submissionViewModel.submitDiagnosisKeys(keys)
+        if (keys.isNotEmpty()) { submissionViewModel.submitDiagnosisKeys(keys)
+        } else { showSubmitErrorDialog()
+        }
```
That's quite unusual formatting ;)

Fabian-K commented on Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragment.kt:
```diff
@@ -177,7 +187,31 @@ class SubmissionResultPositiveOtherWarningFragment : Fragment(),
+    private fun showSubmitErrorDialog() {
+        val submitErrorDialogDialogInstance = DialogHelper.DialogInstance(
+                requireActivity(),
+                R.string.submission_error_dialog_diagnosis_key_title,
+                R.string.submission_error_dialog_diagnosis_key_paired_body,
+                R.string.submission_error_dialog_diagnosis_key_paired_button_positive,
```
same here, can we make it obvious that this is for the no diagnosis keys case?

ritsam replied:
made the necessary changes

Fabian-K commented on Corona-Warn-App/src/androidTest/java/de/rki/coronawarnapp/ui/submission/SubmissionResultPositiveOtherWarningFragmentTest.kt:
```diff
@@ -0,0 +1,69 @@
+     * to be executed in a scenario where test results are positive
+     */
+    @Test
+    fun performPositiveTestResultTest() {
```
I´m not sure how this test is supposed to work. I´m at least missing some sort of setup of the test case. Also is this a test for the positive test result case or for the special case of "no diagnosis keys"?

ritsam replied:
This is a unit test which looks for positive test result case. and then checks if there are no keys. I have a feeling I'll have to rewrite the test again. There us definitely some misunderstanding from my side of scenario where this condition occures

ritsam replied:
The scenario (submitting diagnosis keys) only works when the result is positive
So the test is successful only if there are no keys and the error dialog is shown.

1. When the result is positive, that card is visible on the HomeFragment
2. By clicking on that card we go to the test page where user can choose to share the keys
3. then SubmissionResultPositiveOtherWarningFragment where our logic is in place to check if the keys are present.
4. upon no keys, error dialog is shown where the test registers a click on the button
5. Redirection to HomeFragment happens 
6. Test checks if redirection has happened successfully.

ritsam replied:
@Fabian-K Integration test has been added that mimics 0 keys while submitting.

AndroidMedaGalaxy committed:
code review changes

AndroidMedaGalaxy committed:
integration test added for No diagnosisKey available:

Test added for No Diagnosis key available

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
No submission keys error sliently handled

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
silently handled no key error submission

1. Removed  unceccesary strings
2. redirection to SubmissionDoneFragment when there are no keys available
3. removed navigation action for SubmissionResultPositiveOtherWarningFragment to MainFragment

AndroidMedaGalaxy committed:
code ruse and test case removed at it was not needed

AndroidMedaGalaxy committed:
removed unused import

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
comment adjusted

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

AndroidMedaGalaxy committed:
Merge branch 'dev' into fix/submission_error_dialog

pwoessner committed:
Merge branch 'dev' into fix/submission_error_dialog
